File: threadsplitter.pl
#!/usr/bin/perl
use strict;
use warnings;
use v5.10;

use File::Slurp;
use Mojo::DOM;

use constant MY_NAME => "Aaron van Geffen";

my ($filename) = shift;
my ($export_dir) = shift;
if (!$filename or !(-f $filename) or !$export_dir) {
	warn "Usage: $0 <path to messages.htm> <directory to export to> \n";
	exit 1;
}

if (!(-d $export_dir)) {
	warn $export_dir, " does not exist -- will now be created.\n";
	mkdir $export_dir;
}

say "Reading file " . $filename . "...";
my $messages = read_file($filename);

say "Parsing DOM structure...";
my $dom = Mojo::DOM->new($messages);

say "Finding threads...";
my $collection = $dom->find('div.thread');

say "Found " . $collection->size . " threads...";
if ($collection->size == 0) {
	say "Nothing to do -- stopping.";
	exit;
}

say "Exporting conversation threads to " . $export_dir . "...";
my %conversations;
my $unnamed_conversations = 0;
for my $conversation ($collection->each) {
	my @conversation_partners = grep(!/@{[MY_NAME]}/, split(', ', $conversation->text));
	my $conversation_name = join(', ', @conversation_partners);

	if (defined($conversations{$conversation_name})) {
		$conversations{$conversation_name}++;
		$conversation_name .= ' - ' . $conversations{$conversation_name};
	} else {
		$conversations{$conversation_name} = 1;
	}

	open SPLITMSG, ">", $export_dir . "/" . $conversation_name . ".htm";
	if (tell(SPLITMSG) == -1) {
		print "Could not open file handle for '", $export_dir . "/" . $conversation_name . ".htm' -- trying '";
		$conversation_name = "unnamed_conversation_" . ++$unnamed_conversations;
		say $conversation_name . ".htm";

		open SPLITMSG, ">", $export_dir . "/" . $conversation_name . ".htm";
		if (tell(SPLITMSG) == -1) {
			say "Could not open file handle for '", $export_dir . "/" . $conversation_name . ".htm' either -- cowardly exiting.";
			exit 2;
		}
	}

	print SPLITMSG '<!DOCTYPE html><meta http-equiv="Content-Type" content="text/html; charset=utf-8">';
	print SPLITMSG '<style type="text/css">body { font-family: sans-serif; } .user { font-weight: bold; } .meta { margin-left: 1em; color: #999 } </style>';
	print SPLITMSG '<p>Participants: ', join(', ', @conversation_partners), '</p>';

	my $messages = $conversation->children->reverse;
	my $next_is_header = 0;
	my $previous;

	for my $message ($messages->each) {
		if ($next_is_header) {
			print SPLITMSG $message->to_string;
			print SPLITMSG $previous->to_string;
			$next_is_header = 0;
		} else {
			$previous = $message;
			$next_is_header = 1;
		}
	}

	close SPLITMSG;
}

File: threadsplitter.py
#!/usr/bin/env python3
import argparse
import os
import sys
import xml.etree.ElementTree as etree


def writeThreadToFile(participants, messages):
    if len(participants) == 0:
        print("    WARNING: no participants in thread -- skipping " + str(len(messages)) + " message(s)")

    # What file are we writing to?
    outputFile = os.path.join(args.outdir, ', '.join(participants) + ".html")

    # New file? Start with basic HTML pleasantries.
    if not os.path.exists(outputFile):
        print("Appending to", outputFile)
        handle = open(outputFile, 'w+')
        handle.write('<!DOCTYPE html><meta http-equiv="Content-Type" content="text/html; charset=utf-8">')
        handle.write('<style type="text/css">body { font-family: sans-serif; } .user { font-weight: bold; } .meta { margin-left: 1em; color: #999 } </style>')
        handle.write('<p>Participants: ' + (', '.join(participants)) + '</p>')

    # Otherwise, append to the existing file.
    elif os.path.isfile(outputFile):
        print("  Writing to", outputFile)
        handle = open(outputFile, 'a+')

    # Append each of the messages in order.
    for (header, body) in messages:
        handle.write(header)
        handle.write(body)

    # End with a horizontal rule.
    handle.write('<hr>')
    handle.close()


def processFile(action=writeThreadToFile):
    file = open(args.messages)
    for event, elem in etree.iterparse(file, events=('end', 'start')):
        # Infer our own name if not provided.
        if elem.tag == 'h1' and args.myname == None:
            args.myname = elem.text

        # Process each of the threads individually.
        if 'class' in elem.attrib and elem.attrib['class'] == 'thread':
            participants, messages = processThread(elem)
            action(participants, messages)

        # Free element after use.
        elem.clear()


def processThread(thread):
    messages = []
    participants = []
    savedHead = savedBody = None

    # First, get all messages in this thread.
    for elem in thread.getchildren():
        # Message headers
        if elem.tag == 'div':
            # Keep track of all participants.
            participant = elem.find("./div/span")
            if participant != None and participant.text != None:
                if not participant.text in participants and not participant.text == args.myname:
                    participants.append(participant.text)

            savedHead = etree.tostring(elem, encoding="unicode")

        # Message bodies
        elif elem.tag == 'p':
            savedBody = etree.tostring(elem, encoding="unicode")

        # Store message header and body.
        if savedHead and savedBody:
            messages.append([savedHead, savedBody])
            savedHead = savedBody = None

    # Reverse thread order.
    messages.reverse()

    return (participants, messages)


# Run on command line if not used as a package.
if __name__ == '__main__':
    # Parse arguments from command line.
    parser = argparse.ArgumentParser(description='Split message threads into their own files')
    parser.add_argument('--messages', dest='messages', required=True,
                        help='HTML file from which to read messages')
    parser.add_argument('--outdir', dest='outdir', required=True,
                        help='Folder to write message threads to')
    parser.add_argument('--myname', dest='myname', required=False, default=None,
                        help='Name to ignore when naming conversations')

    args = parser.parse_args()

    # Create output directory if needed.
    if not os.path.isdir(args.outdir):
        if not os.path.exists(args.outdir):
            print("Output dir " + args.outdir + " does not exist yet, creating...")
            os.makedirs(args.outdir)
        else:
            print("Specified output dir " + args.outdir + " is not a directory...")
            sys.exit(1)

    # Sanity check: does our input file exist?
    if not os.path.isfile(args.messages):
        print("Input file dir " + args.messages + " does not exist yet, creating...")
        sys.exit(2)

    # Process the file.
    processFile()
